soroban-env-host/src/host.rs

```diff
@@ -2377,6 +2384,67 @@ impl VmCallerEnv for Host {
         self.create_contract_with_id_preimage(code, id_preimage)
     }
 
+    // Notes on metering: covered by components
+    fn put_tmp_contract_data(
+        &self,
+        _vmcaller: &mut VmCaller<Host>,
+        k: RawVal,
+        v: RawVal,
+    ) -> Result<Void, HostError> {
+        let key = self.from_host_val(k)?;
```

Comment:
I .. don't think you want to do this conversion from k to host val / ScVal key. That's just needless overhead, you can use k as a key directly.

Reply:
+1 to @graydon. You shouldn't convert `RawVal` to `ScVal`. 
Also, you should just keep `Hash` instead of `[u8; 32]` for better clarity. `MeteredMap` can accept any key type that implements `Compare`. So `(Hash, RawVal)` will work. 
Also, you want to make the key type `Rc<(Hash, RawVal)>`, since `MeteredMap`'s metering is calibrated to small keys comparison. This is why the existing `Storage` has `Rc<LedgerKey>` as the key type.

Reply:
Agree on the rc part, I'll do the update. I also can use hash, I suppose it won't make much difference.

However, I'm not sure about using  `RawVal` for keys here. The semantics of the temp storage are the same as for the regular storage and regular storage uses ledger keys. Are we confident that `RawVal` comparisons are always equivalent to the corresponding `ScVal` comparisons, including UDTs etc?

Reply:
If they are not identical, it is a very serious bug, because we interconvert between them constantly, we use binary search to look up map entries in host objects, and "map is always sorted" is an invariant lots of stuff relies on.

I will not claim this bug is impossible or even that we have enough testing to rule it out in the current codebase. Merely that it is essential that it not ship in the final product already.

Reply:
Ok, if that's the invariant, then it should be ok to use the `RawVal`s. I hope we'll find out soon enough if there is some bug...

Reply:
Updated to use RawVals.